http-client/src/main/java/io/airlift/http/client/spnego/SpnegoAuthentication.java

```diff
@@ -58,72 +65,50 @@
         }
     }
 
-    private final LoginContext loginContext;
-    private final GSSCredential clientCredential;
+    private final File keytab;
+    private final File credentialCache;
     private final String remoteServiceName;
+    private final String principal;
+
+    private final LoadingCache<String, LoginContext> loginContextCache = CacheBuilder.newBuilder()
+            .expireAfterWrite(CACHE_EXPIRE_TIME.roundTo(TimeUnit.HOURS), TimeUnit.HOURS)
+            .removalListener(notification -> {
+                try {
+                    ((LoginContext)notification.getValue()).logout();
+                }
+                catch (LoginException e) {
+                    Throwables.propagate(e);
+                }
+            })
+            .build(new CacheLoader<String, LoginContext>()
+                   {
+                       @Override
+                       public LoginContext load(String key)
+                               throws Exception
+                       {
+                           return loadLoginContext();
+                       }
+                   }
+            );
 
     public SpnegoAuthentication(File keytab, File kerberosConfig, File credentialCache, String principal, String remoteServiceName)
     {
         requireNonNull(kerberosConfig, "kerberosConfig is null");
         requireNonNull(remoteServiceName, "remoteServiceName is null");
 
+        this.keytab = keytab;
+        this.principal = principal;
+        this.credentialCache = credentialCache;
         this.remoteServiceName = remoteServiceName;
-
         System.setProperty("java.security.krb5.conf", kerberosConfig.getAbsolutePath());
-
-        try {
```

Comment:
Did anything change with this "introduce method" refactoring? Please do it in a separate commit